File: go_for_data_engineering/05_design_patterns/01_singleton.go
// Singleton via sync.Once — thread-safe, lazy.
package main

import (
	"fmt"
	"sync"
)

type Config struct {
	Env string
}

var (
	instance *Config
	once     sync.Once
)

func GetConfig() *Config {
	once.Do(func() {
		fmt.Println("initializing config (once)")
		instance = &Config{Env: "prod"}
	})
	return instance
}

func main() {
	a := GetConfig()
	b := GetConfig()
	fmt.Println(a == b, a.Env)
}

File: extra_learning/design_patterns/01_singleton.py
"""Singleton via a metaclass — thread-safe lazy initialization."""
from __future__ import annotations
import threading


class SingletonMeta(type):
    _instances: dict[type, object] = {}
    _lock = threading.Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            if cls not in cls._instances:
                cls._instances[cls] = super().__call__(*args, **kwargs)
        return cls._instances[cls]


class Config(metaclass=SingletonMeta):
    def __init__(self) -> None:
        print("init config (once)")
        self.env = "prod"


if __name__ == "__main__":
    a = Config()
    b = Config()
    print(a is b, a.env)
